File: Python3/61. Rotate List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    
    # a helper function to reverse the linked list
    def reverseList(self, head: ListNode, rotate: int):
        prev = None
        for _ in range(rotate):
            head.next, prev, head = prev, head, head.next
        return prev
    
    def rotateRight(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        # special case 1 and 2
        if not head or not head.next:
            return head
        
        # count the total nodes in the linked list
        total = 0
        cur = head
        while cur:
            cur = cur.next
            total += 1
        
        k %= total
        # special case 3
        if k == 0:
            return head
        
        # reverse the whole list
        head = self.reverseList(head, total)
        
        # get the starting node of the second half
        second = head
        for _ in range(k):
            second = second.next
        
        second = self.reverseList(second, total-k)
        #print("head = ", head)
        #print(second)
        # reverse the first half, and cut off the first node
        # because it's duplicate
        first = self.reverseList(head, k)
        #print("first = ", first)
        
        head = first
        
        #print("head = ", head)
        
        # link the first half and the second half
        while first.next:
            first = first.next
            
        first.next = second
        
        return head
        
        
        

File: Java/61. Rotate List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode rotateRight(ListNode head, int k) {
        if(head == null || head.next == null){
            return head;
        }
        
        int total = 0;
        ListNode cur = head;
        
        while(cur != null){
            cur = cur.next;
            total++;
        }
        
        k %= total;
        
        /* special case */
        if(k == 0){
            return head;
        }
        
        cur = head;
        ListNode newHead = head.next;
        // get the new tail 
        for(int i=1; i<total-k; i++){
            cur = cur.next;
            newHead = newHead.next;
        }
        
        /* cut off the new tail from the new head */
        cur.next = null;
        
        /* record the new head */
        cur = newHead;
        
        while(cur.next != null){
            cur = cur.next;
        }
        
        cur.next = head;
        
        return newHead;
        
        
    }
}